src/FlaUI.WebDriver/Controllers/ElementController.cs

```diff
@@ -178,7 +179,33 @@ public async Task<ActionResult> ElementSendKeys([FromRoute] string sessionId, [F
             {
                 return ElementNotInteractable(elementId);
             }
-            element.AsTextBox().Text = elementSendKeysRequest.Text;
+
+            element.Focus();
+
+            // Warning: Deviation from the spec. https://www.w3.org/TR/webdriver2/#element-send-keys says:
+            //
+            // > Set the text insertion caret using set selection range using current text length for both the start and end parameters.
+            //
+            // In English: "the caret should be placed at the end of the text before sending keys". That doesn't seem to be possible
+            // with UIA, meaning that the text gets inserted at the beginning, which is also WinAppDriver's behavior.
+
+            var inputState = session.InputState;
+            var inputId = Guid.NewGuid().ToString();
+            var source = (KeyInputSource)inputState.CreateInputSource("key");
+
+            inputState.AddInputSource(inputId, source);
+
+            try
+            {
+                await ActionsDispatcher.DispatchActionsForString(session, inputId, source, elementSendKeysRequest.Text);
+            }
+            finally
+            {
+                // This isn't in the spec, but we're going to get an assertion failure if we don't clean up.
```

Comment:
The spec says "Try to [clear the modifier key state](https://www.w3.org/TR/webdriver2/#dfn-clear-the-modifier-key-state) with input state, input id, source, undo actions, and browsing context." as a last step, isn't it?

Reply:
It does, and we're currently doing that [at the end of "dispatch actions for string"](https://github.com/FlaUI/FlaUI.WebDriver/pull/33/files/537e42f1d3e02551a76b899793f0a9451e1d2d98#diff-54557db9cee83470bd92a959168a91b0b75c0ec43072a0f74981be7f831c45b5R96) as specified.

The problem is that "clear the modifier key state" uses the "undo actions" defined in "2. Let undo actions be an empty map." in "dispatch actions for string" and not the "input cancel list".

Reply:
Good point. I see the problem that the shift key may be down after the "dispatch the events for a typeable string" and that it is not added to the list of undo actions, plus that "remove an input source" at the end of element send keys asserts that the input cancel list is empty, which it isn't if we do not release somehow. Can we replace the "clear the modifier state" implementation with releasing all actions for that input source instead of just using the undo actions list? Then we do not follow the spec to the letter (you have correctly raised an issue for that) but we do follow its spirit, namely that "clearing the modifier key state" should reset the input state.

Reply:
Yeah I think the "undo actions" list probably shouldn't be there and the undo actions should be added to the input cancel list. I'll try doing that.

Reply:
Done.